ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/DiffSeq.hs

```diff
@@ -0,0 +1,369 @@
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE MonoLocalBinds #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+{- | Sequences of diffs for ledger tables.
+
+   These diff sequences are an instantiation of a strict finger tree with root
+   measures. The tree/sequence itself contains diffs and slot information, while
+   the root measure is the total sum of all diffs in the sequence. The internal
+   measure is used to keep track of sequence length and maximum slot numbers.
+
+   The diff datatype that we use forms a cancellative monoid, which allows for
+   relatively efficient splitting of finger trees with respect to recomputing
+   measures by means of subtracting diffs using the 'stripPrefix' and
+   'stripSuffix' functions that cancellative monoids provide. Namely, if either
+   the left or right part of the split is small in comparison with the input
+   sequence, then we can subtract the diffs in the smaller part from the root
+   measure of the input to (quickly) compute the root measure of the /other/
+   part of the split. This is much faster than computing the root measures from
+   scratch by doing a linear-time pass over the elements of the split parts, or
+   a logarithmic-time pass over intermediate sums of diffs in case we store
+   cumulative diffs in the nodes of the finger tree.
+
+   === Example of fast splits
+
+   As an analogy, consider this example: we have a sequence of consecutive
+   integer numbers @xs = [1..n]@ where @n@ is large, and we define the root
+   measure of the sequence to be the total sum of these numbers, @rmxs = sum
+   [1..n]@ (we assume @rmxs@ is fully evaluated). Say we split this sequence of
+   integer numbers at the index @2@, then we get /left/ and /right/ parts of the
+   split @ys@ and @zs@ respectively.
+
+   > splitAt 2 xs = (ys, zs) = ([1..2], [3..n])
+
+   How should we compute we the root measure @rmys@ of @ys@? Since @ys@ is
+   small, we can just compute @rmys = sum [1..2]@. How should we compute the
+   root measure @rmzs@ of @zs@? We should not compute @rmzs = sum [3..n]@ in
+   this case, since @n@ is large. Instead, we compute @rmzs = rmxs - rmys@,
+   which evaluates to its result in time that is linear in the length of @ys@,
+   in this case @O(1)@.
+
+   === Why not store sums of diffs in the internal measure instead of the root
+       measure?
+
+   We could also have used the interal measure of the strict finger tree to
+   store intermediate sums of diffs for all subtrees of the node. The subtree
+   rooted at the root of the tree would then store the total sum of diffs.
+   However, we would have now to recompute a possibly logarithmic number of sums
+   of diffs when we split or extend the sequence. Given that in @consensus@ we
+   use the total sum of diffs nearly as often as we split or extend the diff
+   sequence, this proved to be too costly. The single-instance root measure
```

Comment:
Thanks!

Flushes are once every 100 blocks, IIRC, so that's  `2^~7` diffs we split from a sequence of `2^~11`. At first glance, it might sound weird that using the root measure would be quicker, because it has a linear cost -- it needs to recompute the sum of `2^~7` diffs instead of `~7` diffs if we were to use the "regular" fingertree split, which is logarithmic. However, if we take into account the following access pattern:

* A = retrieve the total sum of diffs
* B = snoc a diff to the sequence
* C = split 100 diffs from the left of the sequence

We perform A then B a total of 100 times, and then we perform C once. Repeat 

We wrote a benchmark for this scenario, and it turned out the root measured fingertree was faster in this case. If we look at the complexity of these operations, then IIRC for a regular fingertree:

* A = O(1) amortised
* B = O(1) amortised
* C = O(log 100) amortised (?)

For a root-measured fingertree:

* A = O(1) worst-case
* B = O(1) worst-case
* C = O(100) worst-case

So complexity wise, the root-measured fingertree looks worse, but in practice it was a bit better than the regular fingertree

I'm actually not convinced by this root-measured fingertree approach. I wonder if its complexity outweights the performance benefits. Moreover, the V2 LedgerDB is intended to be the default, and it uses "simple" diffs without cancellative properties

Reply:
Pasted a version of this explanation in the haddocks